lib/plugins/aws/package/compile/functions.js

```diff
@@ -219,9 +218,20 @@ class AwsCompileFunctions {
         ? this.serverless.service.package.deploymentBucket
         : { Ref: 'ServerlessDeploymentBucket' };
 
-      functionResource.Properties.Code.S3Key = `${
-        this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      const serverlessPackagedArtifact = artifactFilePath.split(path.sep).indexOf('.serverless');
+
+      // Check if it's not a custom artifact
+      if (serverlessPackagedArtifact > -1) {
+        functionResource.Properties.Code.S3Key = `${
+          this.serverless.service.package.artifactDirectoryName
+        }/${artifactFilePath
+          .split(path.sep)
+          .splice(-1, `${serverlessPackagedArtifact}`)
+          .join(path.sep)}`;
+      } else {
+        functionResource.Properties.Code.S3Key = `${this.serverless.service.package.artifactDirectoryName}/${artifactFilePath}`;
```

Comment:
I think in this situation we should still account for the fact that `artifactFilePath` can start with a `/` or `\` that should be escaped and for S3 I believe we might need to split on `path.sep`, but join on `/` to be consistent, otherwise we might end up in a situation where deploying the same service from Windows and Linux machine will result in different S3 Keys. @medikoo what is your opinion here?

Reply:
I guess it's a good point. I think we may simply unconditionally replace `\` with `/`

Reply:
Done. `S3Key` format is now consistent on *nix and Windows.

Reply:
Hey @scadu - did you push recent changes? If I'm not mistaken the code is the same as before

Reply:
@pgrzesik Just pushed, sorry for that.

Reply:
No worries @scadu :100: I've just checked the change and I believe it's still going to be problematic when the `artifactFilePath` is an absolute path, as it will create an invalid S3Key e.g. if `artifactFilePath` would be `/path/to/file`, in this case it would result in an `S3Key` with two slashes next to each other. I think we should try to transform absolute paths by stripping the `/` at the beginning on the equivalent on Windows. What do you think?

Reply:
I'll take a look.

Reply:
Thank you :bow:

Reply:
There's still hackery but I couldn't figure out anything more elegant. Trying to get tests working again.
The `path.root` is removed now. Not sure if we would also like to convert Windows path like `some\\path\\to\\artifact.zip` to `some/path/to/artifact.zip` for S3Key.

Reply:
Hello :wave: Thank you @scadu I've posted a small "proposal" and I'd love to hear your opinion on the approach. As for convering paths - I think we should ensure it's always converted to using forward slashes for `S3Key`. :+1: